### ISSUE TITLE
Invert foreground color if bg and fg color is too close for readability.

### ISSUE THREAD

**Issue by BLucky-gh** [kovidgoyal/kitty#5278]
**Is your feature request related to a problem? Please describe.**
Oftentimes I have a problem with some themes/colorschemes where some of the 16 terminal colors are very close in "brightness" to the foreground text color, making text unreadable on colored backgrounds like in the screenshot below
![image](https://user-images.githubusercontent.com/41377306/178952689-7b942e2e-816d-47f8-b9aa-a1f0bc98cf0e.png)
I understand this is mostly a theming issue but it's happening with a bunch of popular themes (catppuccin being one of them) and the solution shouldn't be too hard to implement (it's actually already a thing in vscode) so I decided to submit a feature request.

**Describe the solution you'd like**
VSCode already has a solution for this which is inverting the color of the fg if the background is too similar in color (not sure if it's actually inverting or doing somehting else to make it darker but the point is it makes white text black to make white on white readable)

**Describe alternatives you've considered**
I've considered editing every colorscheme I come across to have darker colors but that means wonky looking themes if I'm not too careful about what shades I'm using, and it's also a lot of effort.

**Additional context**
Here's what the same app looks like in vscode's builtin terminal (the colorscheme is slightly different but the same problem is there too:
<img width="340" alt="image" src="https://user-images.githubusercontent.com/41377306/178962975-3f31a2c8-1b1b-42b9-8dc1-85fe9fe811ac.png">

**Comment by kovidgoyal** [comment]
I very much do not want to have kitty messing with colors the user has
specified. The user might have perfectly good reasons for choosing
whatever colors they did choose.

**Comment by page-down** [comment]
This feature is categorized under Accessibility in VS Code.

https://github.com/xtermjs/xterm.js/blob/4.19.0/src/common/Color.ts#L189-L214

> I understand this is mostly a theming issue ...

Maybe changing the theme color does not solve the problem. For example, when the theme defines Yellow as `#ffcc00` and the program sets fg color to Yellow and bg color to `#ffcc00` when outputting text, it will be indistinguishable.

The solution is to let every program you use know your theme's color scheme, or even better, to use RGB colors and let them enhance the contrast.

**Comment by kovidgoyal** [comment]
Seems pretty wrong to me. What if a program deliberately sets the fg and
bg colors to be similar to fade text or even to hide it temporarily. The
terminal emulator should not interfere as it has no way of guessing the
programs intentions.

**Comment by BLucky-gh** [comment]
While both positions are valid, I think at least allowing it as an optional feature (or idk, exposing enough in the kitten api to let others implement it if that's possible? though iirc kittens have to be called manually so I don't think they could be used to continuously change the output buffer thing) would still be very appreciated. And as for programs that deliberately set fg and bg to the same color:
1) that could be fixed by making the feature optional and toggleable
2) it's very rare to see fading animations in TUIs, if ever, and hiding text by setting fg to bg is imo less intuitive and a bit more work than just not outputting the text in the first place

This isn't the first time a terminal emulator has done what a terminal program has been doing or blurred the line between the emulator and the tty within it (for example kitty having most of the features of tmux, including adding a line at the top of the terminal for the tab bar that looks like part of the terminal output, but is actually part of the emulator that the tty inside has no direct access to), so I don't think adding an optional toggleable feature or the ability to code your own feature via an api of sorts wouldn't be much of an interference to programs running inside kitty since at that point it's the user telling the emulator to interfere, not the terminal emulator interfering on its own accord.

TL;DR: I think it would still be nice to let us get close to shooting ourselves in the foot when we know what risks there are and what problems could arise.

**Comment by page-down** [comment]
> ... exposing enough in the kitten api to let others implement it ...

It seems to me that this is not a good fit for Python "API" and would have performance issues.
If you really need this, check out the code in `kitty/shaders.c`.

https://github.com/kovidgoyal/kitty/commit/c06a03ab9652c0991ae810fc6e3682350786a194

> ... adding an optional toggleable feature ...
> ... the user telling the emulator to interfere ...

This reminds me of the discussion of increasing the contrast of bold fonts.

With this added, the user will not know exactly what the color escape sequence will do. There is also no way for the program author to explicitly opt out of this option.

While I think these accessibility features are useful for some people, perhaps it is more important that kitty render the output of the program correctly.
Ensure that all users are treated equally, and achieve consistent rendering by not providing these options.

(Oh, what about color blind users? Perhaps they would appreciate a similar feature? It is very unfortunate that everyone in this world has different health conditions, and not every program can do the right thing. It is not kitty's responsibility to fix their problems though.)

**Comment by BLucky-gh** [comment]
That is a valid concern, thanks for mentioning shaders.c, I might just code my own personal fork for that, shouldn't be too much code.
